### PR TITLE
Fix: bug when removing dynamic sections

### PR DESCRIPTION
**Pull request by NSElvis**
Removing sections isn't working in all cases, I added a few failing tests to reproduce the bug.

Most of the bugs are mainly located in `resetDynamicSectionsWithDictionary` and evaluating which sections where removed compared to your initial values.

### CODE REVIEW

File: Source/FORMDataSource.m [hunk 1]
```diff
@@ -939,9 +948,9 @@ - (void)insertDynamicSectionsForValues:(NSDictionary *)values
     [JSONAttributes enumerateKeysAndObjectsUsingBlock:^(NSString *key, id obj, BOOL *stop) {
         if ([obj isKindOfClass:[NSArray class]]) {
             FORMSection *section = [self sectionWithID:key];
-            [self.formsManager insertTemplateSectionWithID:key
-                                        intoCollectionView:self.collectionView
-                                                 usingForm:section.form];
+            for (NSInteger numberOfSections = 0; numberOfSections < [obj count]; numberOfSections++) {
+                [self.formsManager insertTemplateSectionWithID:key intoCollectionView:self.collectionView usingForm:section.form];
```
Comment (zenangst) [inline Source/FORMDataSource.m]: Shouldn't this conform to our style guide?

https://github.com/hyperoslo/iOS-playbook/blob/master/STYLE_GUIDELINES.md

Comment (NSElvis) [inline Source/FORMDataSource.m]: Yes, should be pushed in the next commit

File: Source/FORMData.h [hunk 2]
```diff
@@ -39,7 +39,7 @@
 - (void)fieldWithID:(NSString *)fieldID includingHiddenFields:(BOOL)includingHiddenFields
          completion:(void (^)(FORMField *field, NSIndexPath *indexPath))completion;
 
-- (void)removeSection:(FORMSection *)removedSection;
+- (void)removeSection:(FORMSection *)removedSection inCollectionView:(UICollectionView *)collectionView;
```
Comment (zenangst) [inline Source/FORMData.h]:
```
- (void)removeSection:(FORMSection *)removedSection 
         inCollectionView:(UICollectionView *)collectionView;
```

File: Source/FORMData.m [hunk 3]
```diff
@@ -493,7 +493,7 @@ - (void)fieldWithID:(NSString *)fieldID includingHiddenFields:(BOOL)includingHid
     if (completion) completion(foundField, indexPath);
 }
 
-- (void)removeSection:(FORMSection *)removedSection
+- (void)removeSection:(FORMSection *)removedSection inCollectionView:(UICollectionView *)collectionView
```
Comment (zenangst) [inline Source/FORMData.m]:
```
- (void)removeSection:(FORMSection *)removedSection 
     inCollectionView:(UICollectionView *)collectionView;
```

File: Source/FORMDataSource.m [hunk 4]
```diff
@@ -443,32 +443,42 @@ - (void)reloadWithDictionary:(NSDictionary *)dictionary
...
         HYPParsedRelationship *parsed = [key hyp_parseRelationship];
-        parsed.attribute = nil;
-        FORMSection *section = [self sectionWithID:[parsed key]];
-        if (section) {
-            [self.formsManager removeSection:section];
+        if (parsed.toMany) {
+            parsed.attribute = nil;
+
+            if (![currentSections containsObject:[parsed key]]) {
+                if (![dictionary[key] isKindOfClass:[NSNull class]]) {
```
Comment (zenangst) [inline Source/FORMDataSource.m]: Could we combine these to if statements into to a boolean conditional ?

File: Source/FORMDataSource.m [hunk 5]
```diff
@@ -594,13 +604,8 @@ - (void)fieldCell:(UICollectionViewCell *)fieldCell updatedWithField:(FORMField
...
             NSString *sectionID = [parsed key];
-            [self.formsManager sectionWithID:sectionID completion:^(FORMSection *section, NSArray *indexPaths) {
-                [self.formsManager removeSection:section];
-
-                if (indexPaths) {
-                    [self.collectionView deleteItemsAtIndexPaths:indexPaths];
-                }
-            }];
+            FORMSection *section = [self.formsManager sectionWithID:sectionID];
+            [self.formsManager removeSection:section inCollectionView:self.collectionView];
```
Comment (zenangst) [inline Source/FORMDataSource.m]:
```
[self.formsManager removeSection:section 
                inCollectionView:self.collectionView];
```

File: Source/FORMData.m [hunk 6]
```diff
@@ -564,9 +569,14 @@ - (void)removeSection:(FORMSection *)removedSection
 
     NSString *sectionID = removedSection.sectionID;
     FORMGroup *form = removedSection.form;
-    [form.sections removeObject:removedSection];
-    relationshipIndex = 0;
+    [self sectionWithID:sectionID
+             completion:^(FORMSection *section, NSArray *indexPaths) {
+                 FORMGroup *f = section.form;
```
Comment (zenangst) [inline Source/FORMData.m]: Why just `f` and not `form` ?

Comment (NSElvis) [inline Source/FORMData.m]: hahaha, forgot about that!!!!

**Comment by zenangst** [comment]
Looks good! :star:

**Pull request merged**